File: cppStudy/solved/1005 ACM Craft.cpp
#include <bits/stdc++.h>
using namespace std;
// 1005 ACM Craft
#define FOR(i, n, m) for(int i=(n); i<(m); i++)
constexpr int LMT = 1002;


int main(void) {
    ios::sync_with_stdio(0);
    cin.tie(0);

    int tc;
    cin >> tc;
    FOR(t, 0, tc) {
        int tC[LMT] = {};
        vector<int> preQ[LMT];
        vector<int> proQ[LMT];
        queue<int> noDeg;
        int dp[LMT] = {};
        int inDeg[LMT] = {};
        int n, m;
        cin >> n >> m;
        FOR(i, 1, n+1) cin >> tC[i];
        FOR(i, 0, m) {
            int a, b;
            cin >> a >> b;
            proQ[a].push_back(b);
            preQ[b].push_back(a);
            inDeg[b] ++;
        }
        int tgt;
        cin >> tgt;
        FOR(i, 1, n + 1) if (!inDeg[i]) noDeg.push(i);
        while (1) {
            int v = noDeg.front(); noDeg.pop();
            if (preQ[v].empty()) dp[v] = tC[v];
            else {
                int mxV = 0;
                for (auto pV : preQ[v]) mxV = max(mxV, dp[pV]);
                dp[v] = mxV + tC[v];
            }
            if (v == tgt) break;
            for (auto nv : proQ[v]) {
                inDeg[nv] -= 1;
                if (!inDeg[nv]) noDeg.push(nv);
            }
        }
        cout << dp[tgt] << '\n';
    }
}

File: python/Solved/BOJ/1005 ACM Craft.py
# 1005 ACM Craft
input = __import__('sys').stdin.readline

for _ in range(int(input())):
    n, m = map(int, input().split())
    tC = [0] + list(map(int, input().split()))
    dp = [0]*(n+1)
    preQ = [[] for _ in range(n+1)]
    proQ = [[] for _ in range(n+1)]
    inDeg = [0]*(n+1)
    noDeg = []
    for i in range(m):
        a, b = map(int, input().split())
        proQ[a].append(b)
        preQ[b].append(a)
        inDeg[b] += 1
    target = int(input())
    for i in range(1, n+1):
        if not inDeg[i]: noDeg.append(i)
    while noDeg:
        v = noDeg.pop()
        if preQ[v]: dp[v] = max([dp[i] for i in preQ[v]]) + tC[v]
        else: dp[v] = tC[v]
        if v == target: break
        for nv in proQ[v]:
            inDeg[nv] -= 1
            if not inDeg[nv]: noDeg.append(nv)
    print(dp[target])